### DISCUSSION TITLE
How to get HMR to work behind a reverse SSL proxy?

### DISCUSSION THREAD

**Discussion by HendrikJan** [nuxt/nuxt#16377] [Questions]
Hi,

I'm using Nuxt without HTTPS and then use an Apache reverse proxy to make the website use HTTPS.
(this includes websockets that get translated from `ws:` to `wss` using proxy_wstunnel_module)
Now when trying to use Nuxt-3 with Vite, I cannot get the hot-reload to work.

The problem is:
- Nuxt thinks it is working on HTTP but the browser is actually on HTTPS
- Nuxt tells the browser to open an insecure websocket on ws:/mysite.com/_nuxt_
- Browser does not allow to open insecure websockets from a secure webpage

When setting **vite.server.hmr.protocol** to `ws` I get this error in my browser:
`
Mixed Content: The page at 'https://tapp.docker:9004/report' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://tapp.docker:1304/_nuxt/'. This request has been blocked; this endpoint must be available over WSS.
`

When setting **vite.server.hmr.protocol** to `wss` I get this error and the page keeps reloading:
`
WebSocket connection to 'wss://tapp.docker:1304/_nuxt/' failed: Error in connection establishment: net::ERR_CONNECTION_CLOSED
`

Is there a way to get Nuxt3 to work behind a reverse SSL proxy?

**Comment by tylerforesthauser** [comment]
Unfortunately I'm very rusty with Apache so I can't speak to the specifics of setting _it_ up, but I'll attempt to provide an example. Additionally, I'm not sure how the proxy_wstunnel_module will affect things. But I believe there are two fundamental things you'd need to do:

1. Set up a location in Apache, perhaps something like this:
```apache
<Location /_nuxt/hmr/>
  ProxyPass "http://127.0.0.1:24626"
</Location>
```
2. Then you'd want to ensure you've got your Vite server hmr settings configured as follows:
```js
export default defineNuxtConfig({
  vite: {
    server: {
      hmr: {
        protocol: 'wss',
        clientPort: 443,
        port: 24626,
        path: 'hmr/',
      },
    },
  },
})
```

Just ensure your port numbers match up (24626, in my case). 

Hopefully this is of some use!

**Reply by HendrikJan** [reply]
Your solution seems to fix the problem.

The Apache config seems to be working now, but I run into a lot of other problems so I cannot be sure yet.
If I'm able to fix my other problems, I'll post some details about the Apache config changes I made.

**Reply by MathieuB1** [reply]
To complete the above answer when working with apache + docker + nuxt3:

1. Add this conf to nuxt3

nuxt.config.ts
```
export default defineNuxtConfig({
    vite: {
        server: {
          hmr: {
            protocol: 'wss',
            clientPort: 443,
            port: 24678,
            path: 'hmr/'
          }
        }
    }
})
```

2. activate apache `proxy_wstunnel`  module:

`a2enmod ssl proxy proxy_http proxy_wstunnel headers`

3. set this apache conf (the order matter here):

```
    # Nuxt3 for Hot Reload in Dev (Websockets)
    ProxyPass /_nuxt/hmr/ ws://mycontainername:24678/_nuxt/hmr/

    # Nuxt3 Conf
    ProxyPass /_nuxt http://mycontainername:3000/_nuxt
    ProxyPassReverse /_nuxt http://mycontainername:3000/_nuxt
```
